File: web/src/sync/digest.ts
import type { CollectionName } from "./registry";
import type { SyncDatabase } from "./store";

/**
 * The local half of the daily divergence check.
 *
 * **It hashes exactly what the server would have sent**, in exactly the same
 * order and the same string form: `${id}:${updated_at}\n` per document, ordered
 * by `(updated_at, id)`. The server computes its side in Python over the same
 * strings its pull serialises, for this reason — a SQL `md5(string_agg(...))`
 * would compare a timestamp Postgres formatted against one the browser stored,
 * which is a permanent false mismatch that re-pulls every collection every day.
 */

export interface LocalDigest {
  count: number;
  checksum: string;
}

export async function localDigest(
  database: SyncDatabase,
  name: CollectionName,
): Promise<LocalDigest> {
  const rows = (await database.collections[name]!.find().exec()) as unknown as {
    id: string;
    updated_at: string;
  }[];
  const ordered = rows
    .map((row) => ({ id: row.id, updated_at: row.updated_at }))
    .sort((a, b) =>
      a.updated_at === b.updated_at
        ? a.id.localeCompare(b.id)
        : a.updated_at.localeCompare(b.updated_at),
    );
  const body = ordered.map((row) => `${row.id}:${row.updated_at}\n`).join("");
  return { count: ordered.length, checksum: await sha256(body) };
}

async function sha256(value: string): Promise<string> {
  const bytes = new TextEncoder().encode(value);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return [...new Uint8Array(digest)]
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

File: core/sync/digest.py
"""`GET /api/sync/digest` -- the daily divergence check.

The safety horizon is an engineering choice and not a bet **because of this
endpoint**. It answers, per collection, the row count and a checksum over
`(id, updated_at)` for this device's predicate at the horizon. The client
compares it against its local store once a day on first idle; a mismatch resets
that collection's checkpoint and re-pulls it from zero.

That turns two otherwise-silent permanent losses into a one-day-late repair: a
row written inside a transaction that outlived the horizon, and a `customers`
row that aged out of the recency window behind the device's own cursor and could
therefore never be served as a departure.

**The checksum is computed in Python, over the same strings `pull.py` would have
sent.** A SQL `md5(string_agg(...))` would be faster and would compare a
timestamp Postgres formatted against one the browser stored, which is a
permanent false mismatch that re-pulls every collection every day. One
formatter, one hash, both sides.
"""

import hashlib

from core.sync import registry


def collection_digest(collection, *, tenant_id, location_id, cursor_limit, options):
    """`(count, checksum)` for one collection at the horizon."""
    rows = (
        collection.base(tenant_id, location_id, options)
        .filter(collection.member_q(options), updated_at__lte=cursor_limit)
        .order_by("updated_at", "id")
        .values("id", "updated_at")
    )
    digest = hashlib.sha256()
    count = 0
    for record in rows.iterator(chunk_size=2000):
        head = registry.head(record)
        digest.update(f"{head['id']}:{head['updated_at']}\n".encode("utf-8"))
        count += 1
    return count, digest.hexdigest()


def build(*, tenant_id, location_id, cursor_limit, options):
    """Every collection's digest, in registry order."""
    answer = {}
    for collection in registry.COLLECTIONS:
        count, checksum = collection_digest(
            collection,
            tenant_id=tenant_id,
            location_id=location_id,
            cursor_limit=cursor_limit,
            options=options,
        )
        answer[collection.name] = {"count": count, "checksum": checksum}
    return answer
